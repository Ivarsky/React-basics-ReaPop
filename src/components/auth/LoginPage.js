import { useContext, useState } from 'react';
import Button from '../shared/Button';
import { login } from './service';
import { useLocation, useNavigate } from 'react-router-dom';
import { AuthContext } from './context';

// eslint-disable-next-line react/prop-types
function LoginPage() {
  const { onLogin } = useContext(AuthContext);
  const navigate = useNavigate();
  const location = useLocation();

  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(null);
  // eslint-disable-next-line no-undef
  const [credentials, setCredentials] = useState({
    username: '',
    password: '',
  });

  const resetError = () => {
    setError(null);
  };

  const handleSubmit = async event => {
    event.preventDefault();

    resetError(), setIsLoading(true);
    try {
      await login(credentials);
      setIsLoading(false);
      //Logged In
      onLogin();
      //Redirect to ultima pagina o home
      const to = location.state?.from?.pathname || '/';
      navigate(to);
    } catch (error) {
      setError(error);
      setIsLoading(false);
    }
  };

  const handleChange = event => {
    console.log(event.target.value, event.target.name);

    // if (event.target.name === 'username'){
    //   setCredentials({...credentials, username: event.target.value});
    // }
    // if (event.target.name === 'password') {
    //   setCredentials({...credentials, password: event.target.value});
    // }
    setCredentials({ ...credentials, [event.target.name]: event.target.value });
  };

  const buttonDisabled =
    isLoading || !credentials.username || !credentials.password;

  return (
    <div>
      <h1>Log in to your Account</h1>
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="username"
          onChange={handleChange}
          value={credentials.username}
        />
        <input
          type="password"
          name="password"
          onChange={handleChange}
          value={credentials.password}
        />
        <Button type="submit" variant="primary" disabled={buttonDisabled}>
          Log In
        </Button>
      </form>
      {error && (
        <div onClick={resetError} className="loginPage-error">
          {error.message}
        </div>
      )}
    </div>
  );
}

export default LoginPage;
