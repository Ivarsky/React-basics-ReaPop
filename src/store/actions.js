import {
  ADVERTS_LOADED,
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCES,
  AUTH_LOGOUT,
  UI_RESET_ERROR,
} from './types';

export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST,
});

export const authLoginSucces = () => ({
  type: AUTH_LOGIN_SUCCES,
});

export const authLoginFailure = error => ({
  type: AUTH_LOGIN_FAILURE,
  error: true,
  payload: error,
});

export const authLogout = () => ({
  type: AUTH_LOGOUT,
});

export const advertsLoaded = adverts => ({
  type: ADVERTS_LOADED,
  payload: adverts,
});

export const uiResetError = () => ({
  type: UI_RESET_ERROR,
});
