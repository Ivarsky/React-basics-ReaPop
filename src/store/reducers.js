/* eslint-disable */
import {
  ADVERTS_LOADED,
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCES,
  AUTH_LOGOUT,
  UI_RESET_ERROR,
} from './types';

export const defaultState = {
  auth: false,
  adverts: [],
  ui: {
    isLoading: false,
    error: null,
  },
};

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_SUCCES:
      return true;

    case AUTH_LOGOUT:
      return false;

    default:
      return state;
  }
}

export function adverts(state = defaultState.adverts, action) {
  if (action.type === ADVERTS_LOADED) {
    return action.payload;
  }
  return state;
}

export function ui(state = defaultState.ui, action) {
  switch (action.type) {
    case AUTH_LOGIN_REQUEST:
      return { isLoading: true, error: null };

    case AUTH_LOGIN_FAILURE:
      return { isLoading: false, error: action.payload };

    case AUTH_LOGIN_SUCCES:
      return { isLoading: false, error: null };

    case UI_RESET_ERROR:
      return { ...state, error: null };

    default:
      return state;
  }
}
