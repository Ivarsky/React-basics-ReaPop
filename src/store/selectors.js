/* eslint-disable no-undef */
/* eslint-disable no-debugger */
export const getIsLogged = (state) => state.auth;

export const getAdverts = (state) =>
  state.adverts.areLoaded ? state.adverts.data : [];

//TODO: borra esto
//export const getAdvert = (state, advertId) =>
//  getAdverts(state).find(advert => advert.id === advertId);

export const getAdvert = (advertId) => (state) =>
  state.adverts.data.find((advert) => advert.id === advertId);

export const getUi = (state) => state.ui;

export const areAdvertsLoaded = (state) => state.adverts.areLoaded;
