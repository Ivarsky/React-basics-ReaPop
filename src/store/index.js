import { createStore, combineReducers, applyMiddleware } from "redux";
import { composeWithDevTools } from "@redux-devtools/extension";
import thunk from "redux-thunk";
import * as auth from "../components/auth/service";
import * as adverts from "../components/adverts/service";
import * as reducers from "./reducers";
import * as actionCreators from "./actions";

const reducer = combineReducers(reducers);

const componseEnhancers = composeWithDevTools({
  actionCreators,
});

export default function configureStore(preloadedState, { router }) {
  const middleware = [
    thunk.withExtraArgument({ service: { auth, adverts }, router }),
  ];

  const store = createStore(
    reducer,
    preloadedState,
    componseEnhancers(applyMiddleware(...middleware))
  );
  return store;
}
