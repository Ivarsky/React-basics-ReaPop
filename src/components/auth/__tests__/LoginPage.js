import { fireEvent, render, screen } from "@testing-library/react";
import LoginPage from "../LoginPage";
import { Provider } from "react-redux";
import { defaultState } from "../../../store/reducers";
import { authLogin, uiResetError } from "../../../store/actions";

jest.mock("../../../store/actions");

const errorMessage = "Unauthorized, click to dismiss";

const defaultStateWithError = {
  ...defaultState,
  ui: {
    ...defaultState.ui,
    error: { message: "Unauthorized" },
  },
};

describe("LoginPage", () => {
  const renderComponent = () => {
    const store = {
      getState: () => defaultState,
      subscribe: () => {},
      dispatch: () => {},
    };
    return render(
      <Provider store={store}>
        <LoginPage />
      </Provider>
    );
  };

  const renderComponentWithError = () => {
    const store = {
      getState: () => defaultStateWithError,
      subscribe: () => {},
      dispatch: () => {},
    };

    return render(
      <Provider store={store}>
        <LoginPage />
      </Provider>
    );
  };

  test("snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test("should dispatch authLogin action", () => {
    const email = "cabesa";
    const password = "1234";
    const checkbox = true;

    renderComponent();

    const emailInput = screen.getByLabelText(/Email address/);
    const passwordInput = screen.getByLabelText(/Password/);
    const submitButton = screen.getByRole("button");
    const checkboxButton = screen.getByRole("checkbox");

    expect(submitButton).toBeDisabled();
    expect(checkboxButton).not.toBeChecked();

    fireEvent.change(emailInput, { target: { value: email } });
    fireEvent.change(passwordInput, { target: { value: password } });
    fireEvent.click(checkboxButton);

    expect(submitButton).toBeEnabled();
    expect(checkboxButton).toBeChecked();

    fireEvent.click(submitButton);

    expect(authLogin).toHaveBeenCalledWith({
      checkbox: checkbox,
      credentials: { email, password },
    });
  });

  test("should display an error", () => {
    renderComponentWithError();

    const errorElement = screen.getByText(errorMessage);
    expect(errorElement).toBeInTheDocument();

    fireEvent.click(errorElement);

    expect(uiResetError).toHaveBeenCalled();
  });
});
