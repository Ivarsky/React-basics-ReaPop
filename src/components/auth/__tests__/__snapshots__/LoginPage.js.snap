// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div>
    <h1>
      Log in to your Account
    </h1>
    <form
      class="mx-auto"
      style="max-width: 500px;"
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="formGroupEmail"
        >
          Email address
        </label>
        <input
          class="form-control"
          id="formGroupEmail"
          name="email"
          placeholder="Enter email"
          type="email"
          value=""
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="formGroupPassword"
        >
          Password
        </label>
        <input
          class="form-control"
          id="formGroupPassword"
          name="password"
          placeholder="Enter Password"
          type="password"
          value=""
        />
      </div>
      <div
        class="d-flex align-items-center justify-content-end"
      >
        <label
          class="form-label"
        >
          Keep me logged in
        </label>
        <div
          class="mx-3 form-switch"
        >
          <input
            class="form-check-input"
            type="checkbox"
          />
        </div>
      </div>
      <button
        class="btn btn-primary"
        disabled=""
        type="Submit"
      >
        Log in
      </button>
    </form>
  </div>
</div>
`;
